052-prometheus-metrics-reporter.md

```diff
@@ -0,0 +1,156 @@
+
+For each type of metrics, Kafka exposes a reporter interface to expose metrics to monitoring systems. KafkaMetrics use [org.apache.kafka.common.metrics.MetricsReporter](https://kafka.apache.org/35/javadoc/org/apache/kafka/common/metrics/MetricsReporter.html) and YammerMetrics use `kafka.metrics.KafkaMetricsReporter` which is not officially part of the public API. Kafka has built-in metrics reporter implementations for JMX for both types.
+
+At the moment Strimzi relies on these default JMX reporters and uses [jmx_exporter](https://github.com/prometheus/jmx_exporter) which is a Java agent that retrieves metrics via JMX and exposes them over an HTTP endpoint in the prometheus format. Then Prometheus is configured to scrape that endpoint to retrieve the Kafka metrics.
+
+![Current Situation](images/052-current.png)
+
+- `org.apache.kafka.common.metrics.JmxReporter` is the reporter implementation for the Kafka Metrics.
+- `org.apache.kafka.server.metrics.FilteringJmxReporter` is the reporter implementation for the Yammer Metrics. It’s named `FilteringJmxReporter` because it extends the JmxReporter class from the Yammer library and adds an option to select the metrics to report.
+
+
+## Motivation
+
+I propose updating the metrics collection pipeline for the following reasons:
+
+1. The current metrics collection pipeline is pretty convoluted. We have metrics reporters first exposing metrics via JMX before using a Java agent to expose them again via HTTP. Using metrics reporters to directly expose metrics to Prometheus would significantly simplify it by removing JMX from the picture and removing jmx_exporter.
+1. Each component along the pipeline has its own configurations and specificities. For example, since Kafka 3.4.0, it’s possible to disable the JmxReporter for KafkaMetrics by setting [auto.include.jmx.reporter](https://kafka.apache.org/documentation/#brokerconfigs_auto.include.jmx.reporter) to `false`. It is currently not possible to disable `FilteringJmxReporter` (it’s something I plan to address in the near future). Also both jmx_exporter and the built-in reporter allow selecting metrics to collect. 
+1. The jmx_exporter Java agent supports complex metrics mapping rules. These rules allow renaming metrics and it can make it hard to investigate issues as metrics could have different names. When metrics are renamed, the grafana dashboards we provide can’t be reused. In addition, due to the rules, it performs badly when there’s a very large number of metrics due to a lot of topic/partitions. 
+1. Recently jmx_exporter was impacted by CVEs in its dependencies (CVE-2022-38752, CVE-2022-25857).
+
+
+## Proposal
+
+The proposal is to build a metrics reporter, `PrometheusMetricsReporter`, that implements both `org.apache.kafka.common.metrics.MetricsReporter` and `kafka.metrics.KafkaMetricsReporter`, and that directly exposes metrics via an HTTP endpoint in the Prometheus format. This will be a new project/repository under the Strimzi organization.
```

Comment:
Is it necessary for it to implement both? From Jakub's point avoid I guess we don't need to expose class names in the CR. A reason to not implement both is: Should the Kafka metric plugin interfaces change/get replaced then you might end at that point not be able to implement all of them (e.g. method collisions). What advantages come from having a single class implementing both?

Reply:
We can have 2 classes each implementing one interface. I agree the reporter class names don't need to be exposed to users.